HDDS-3161. Block illegal characters when creating keys.

cku328 opened the pull request:
## What changes were proposed in this pull request?

Added illegal character check (using regular expressions) for key name when creating key.

For the definition of illegal characters, I refer to [Amazon S3's object key naming guide](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html), which specifies the characters to avoid.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3161

## How was this patch tested?

- Ran UTs.
- Using Ozone Shell to create a key with illegal characters in the name, errors can be detected, like this: 
![test](https://user-images.githubusercontent.com/14295594/79052750-358d0280-7c6b-11ea-9282-f2b8bf703dff.png)

elek commented:
Thanks @cku328 to create this patch.  I am fine with following the S3 naming convention but not sure how will it work when buckets will be accessed from NFS/ozone fs. 

What do you think @arp7 , can we do this restriction?

From technical point of view: Didn't check the patch very closely, but it seems to be a client side check. I think it should be checked on the server side, too (but it might be included)...

mukul1987 commented on hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/client/HddsClientUtils.java:
```diff
@@ -196,6 +196,27 @@ public static void verifyResourceName(String... resourceNames) {
+    if (keyName == null) {
+      throw new IllegalArgumentException("Key name is null");
+    }
+    String regex = "^[^^{}<>^?%~#`\\[\\]\\|\\\\(\\x80-\\xff)]$";
```
Please move this regex to OzoneConsts and please add a simplified example on what characters are not allowed.

cku328 replied:
Thanks, I've moved the regex to OzoneConsts and explained it.

cku328 commented:
Thanks to @elek for the comments and @mukul1987 for the review.

I referenced the commit and review comments for this pull request, implemented the key name check at `OMKeyCreateRequest#preExecute`, and applied it the same way to other requests (e.g.: CreateFile, RenameKey). 

In the newly submitted patch, I fixed some of the requests for missing check names, and I tested the put file through OzoneFS and it worked.

![test](https://user-images.githubusercontent.com/14295594/79781045-7d98dd00-836f-11ea-9646-a1345f41903d.png)

adoroszlai commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java:
```diff
@@ -328,5 +328,19 @@ private OzoneConsts() {
+   * and Non-printable ASCII characters (128–255 decimal characters).
+   * https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html
+   */
+  public static final String KEYNAME_AVOID_CHARACTERS_REGEX =
+          "^[^^{}<>^?%~#`\\[\\]\\|\\\\(\\x80-\\xff)]$";
```
Constants for regex patterns should be pre-compiled (`Pattern.compile(...)`).

cku328 replied:
Thanks! I fixed it.

cku328 commented on hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/client/HddsClientUtils.java:
```diff
@@ -196,6 +196,27 @@ public static void verifyResourceName(String... resourceNames) {
+    for (int index = 0; index < keyName.length(); index++) {
+      char currChar = keyName.charAt(index);
+      if (!(Character.toString(currChar)
+              .matches(OzoneConsts.KEYNAME_AVOID_CHARACTERS_REGEX))){
```
Thanks @adoroszlai for the comment.
Checking each character independently with regular expressions to know exactly which one is not allowed. But I feel the same way, independently check each character is expensive, so I modified it to use full-string to check.

cku328 commented:
Thanks @adoroszlai for the review.
In the new patch, this keyname check is optional and it can be enabled by Ozone global config (default is false).

In `ozone-site.xml`, add the following snippet to enable checking:
```
<property>
  <name>ozone.om.keyname.character.check.enabled</name>
  <value>true</value>
</property>
```

adoroszlai commented on hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/client/HddsClientUtils.java:
```diff
@@ -196,6 +196,24 @@ public static void verifyResourceName(String... resourceNames) {
+    }
+    if(!OzoneConsts.KEYNAME_ILLEGAL_CHARACTER_CHECK_REGEX
+            .matcher(keyName).matches()){
+      throw new IllegalArgumentException("key name contains " +
+              "illegal characters.");
```
I would like to propose making this message consistent with the one in `OmUtils`, so that `ozone fs -put` and `ozone sh key put` both print the same error.  I think it can be most simply done by moving the message from `OmUtils` here and letting the `OMException` propagate the underlying exception's message.

```suggestion
      throw new IllegalArgumentException("Invalid key name: " + keyName);
```

If you agree, please feel free to wait for others' reviews before making this change, to reduce the number of rounds.

cku328 replied:
Thanks for the suggestion, I agree with you. 
I'll wait for others' reviews and suggestions before I make this change.

adoroszlai commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java:
```diff
@@ -506,4 +506,17 @@ public static long getOMClientRpcTimeOut(Configuration configuration) {
+    try {
+      HddsClientUtils.verifyKeyName(keyName);
+    } catch (IllegalArgumentException e) {
+      throw new OMException("Invalid key name: " + keyName,
```
Corresponding to the suggestion for `HddsClientUtils`:

```suggestion
      throw new OMException(e.getMessage(),
```

cku328 commented:
Update message of exception and resolve conflict. (latest master-branch #399)

elek commented:
This patch seems to be reviewed and all the comments are addressed. Will merge it if no objections.

I resolved the conflicts and started a new build to get a green build with the latest master.

elek commented:
:musical_note:  Green build, green build, tada-tada-tada... :musical_note: Merging it right now, tada-tada-tada :musical_note: 

Thanks the patch @cku328 (and sorry for the long review time) and the review @adoroszlai
